client.go

```diff
@@ -1,219 +1,101 @@
+	c := Client{conn: conn}
 
-	go c.runLoop()
+	sess := newSession(conn)
+	err := sess.Connect()
+	if err != nil {
+		return c, err
+	}
 
-	return c
+	return Client{conn, sess, sync.Mutex{}}, nil
 }
 
-func NewClientWithCert(gw string, cert tls.Certificate) Client {
+// NewClientWithCert creates a client of the Apple gateway given a certificate.
+func NewClientWithCert(gw string, cert tls.Certificate) (Client, error) {
 	conn := NewConnWithCert(gw, cert)
-
-	return newClientWithConn(gw, conn)
+	return newClientWithConn(conn)
 }
 
+// NewClient is a helper that accepts a certificate/key pair.
 func NewClient(gw string, cert string, key string) (Client, error) {
 	conn, err := NewConn(gw, cert, key)
 	if err != nil {
 		return Client{}, err
 	}
 
-	return newClientWithConn(gw, conn), nil
+	return newClientWithConn(conn)
 }
 
+// NewClientWithFiles is a helper that loads a certificate/key from files.
 func NewClientWithFiles(gw string, certFile string, keyFile string) (Client, error) {
 	conn, err := NewConnWithFiles(gw, certFile, keyFile)
 	if err != nil {
 		return Client{}, err
 	}
 
-	return newClientWithConn(gw, conn), nil
+	return newClientWithConn(conn)
 }
 
+// Send a notification, handling disconnections.
 func (c *Client) Send(n Notification) error {
-	c.notifs <- n
-	return nil
-}
-
-func (c *Client) reportFailedPush(v interface{}, err *Error) {
-	failedNotif, ok := v.(Notification)
-	if !ok || v == nil {
-		return
+	if c.sess.Disconnected() {
+		c.reconnectAndRequeue()
 	}
 
-	select {
-	case c.FailedNotifs <- NotificationResult{Notif: failedNotif, Err: *err}:
-	default:
-	}
+	return c.sess.Send(n)
 }
 
-func (c *Client) requeue(cursor *list.Element) {
-	// If `cursor` is not nil, this means there are notifications that
-	// need to be delivered (or redelivered)
-	for ; cursor != nil; cursor = cursor.Next() {
-		if n, ok := cursor.Value.(Notification); ok {
-			go func() { c.notifs <- n }()
-		}
-	}
-}
-
-func (c *Client) handleError(err *Error, buffer *buffer) *list.Element {
-	cursor := buffer.Back()
+func (c *Client) reconnectAndRequeue() {
+	c.sessm.Lock()
+	defer c.sessm.Unlock()
 
-	for cursor != nil {
-		// Get notification
-		n, _ := cursor.Value.(Notification)
+	// Pull off undelivered notifications
+	notifs := c.sess.RequeueableNotifications()
 
-		// If the notification, move cursor after the trouble notification
-		if n.Identifier == err.Identifier {
-			go c.reportFailedPush(cursor.Value, err)
+	c.reconnect()
 
-			next := cursor.Next()
-
-			buffer.Remove(cursor)
-			return next
-		}
-
-		cursor = cursor.Prev()
+	for _, n := range notifs {
+		// TODO handle error from sending
+		c.sess.Send(n)
 	}
-
-	return cursor
 }
 
-func (c *Client) runLoop() {
-	sent := newBuffer(50)
-	cursor := sent.Front()
+func (c *Client) reconnect() {
+	c.sess = nil
 
-	// APNS connection
-	for {
-		err := c.Conn.Connect()
+	for c.sess == nil {
+		sess := newSession(c.conn)
+
+		err := sess.Connect()
```

Comment:
It looks like sess.Connect() just handles the state machine. This doesn't actually create a new connection. Should it be calling NewConn()?